### ISSUE TITLE
Segfault or infinite loop after changing a regexp

### ISSUE THREAD

**Issue by pskocik** [skvadrik/re2c#166]
I thought it would be fun to automatically linkify C standard references so used re2c with approximately this skeleton:

```
typedef struct input_tp 
{ char *buf, *region, *tok, *yycursor, *yymarker; } 
input_tp;

/*!re2c
        re2c:define:YYCTYPE = char;
        re2c:define:YYCURSOR = X->yycursor;
        re2c:define:YYMARKER = X->yymarker;
        re2c:yyfill:enable = 0;

        Z = "\x00";
        StdRef = [0-9]+('.'[0-9]+)+;
*/

int lex(input_tp* X)
{ 
    for (;;) {
    X->tok = X->yycursor;
    /*!re2c
        StdRef { }
        Z   { return 0; }
        *   { cont; }
    */
    } 

    return 0;
}

int main(int argc, char** argv)
{
    input_tp X;
    X.region=X.tok=X.yycursor=X.buf = argv[1];
    lex(&X);
    return 0;
}

```

This worked well, but then I tried to match references that mention paragraphs too:

`StdRef = [0-9]+('.'[0-9]+)+([/pP][0-9]+)?;

and the added ([/pP][0-9]+)? started causing the code to segfault and replacing it with ([/pP][0-9]*)? made it go into an infinite loop.

Am I doing something wrong?

`

**Comment by skvadrik** [comment]
Empty action in `StdRef` rule looks bad:
```
        StdRef { }
```
You should add something like `continue;` or `return 0;`, otherwise control flow will just proceed to whatever happens to be there. Re2c pastes user-defined actions into the generated file as-is, it does not add any `goto end` statements.

Look at the generated code and you'll probably understand what happens.

**Comment by pskocik** [comment]
Thanks In this context the empty { } was a transcription error. However I did store just the pointer to the buffer and then tried to match again later with just /!re2c  StdRef { } */ (because I knew there would be a match there) and it was the empty braces in the rematch that caused the infinite loop with the changed regex (with the simpler one it worked). I changed it to a goto jump (because on the rematch I was no longer in a parsing loop). Apologies for publishing my stupidities here. re2c is absolutely amazing.
